rfc8252_http_server/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+            }
+            .render()
+            .unwrap(),
+        ))
+}
+
+async fn threaded_error(
+    oidc_params_sender: mpsc::Sender<StdResult<OidcResponseParams, String>>,
+    error: &str,
+    error_description: &str,
+) -> Result<HttpResponse> {
+    let _ = oidc_params_sender.send(Err(format!("{}: {}", error, error_description)));
+    Ok(HttpResponse::build(http::StatusCode::BAD_REQUEST)
+        .content_type("text/html; charset=utf-8")
+        .body(
+            OIDCErrorPage {
+                product_docs_link: "https://www.mongodb.com/docs/atlas/data-federation/query/sql/drivers/odbc/connect",
+                product_docs_name: "Atlas SQL ODBC Driver",
+                error,
+                error_uri: "https://www.mongodb.com/docs/atlas/data-federation/query/sql/drivers/odbc/connect/bad_oidc_login",
+                error_description,
+            }
+            .render()
+            .unwrap(),
+        ))
+}
+
+async fn tokio_error(
+    oidc_params_sender: tokio_mpsc::Sender<StdResult<OidcResponseParams, String>>,
+    error: &str,
+    error_description: &str,
+) -> Result<HttpResponse> {
+    let _ = oidc_params_sender
+        .send(Err(format!("{}: {}", error, error_description)))
+        .await;
+    Ok(HttpResponse::build(http::StatusCode::BAD_REQUEST)
+        .content_type("text/html; charset=utf-8")
+        .body(
+            OIDCErrorPage {
+                product_docs_link: "https://www.mongodb.com/docs/atlas/data-federation/query/sql/drivers/odbc/connect",
+                product_docs_name: "Atlas SQL ODBC Driver",
+                error,
+                error_uri: "https://www.mongodb.com/docs/atlas/data-federation/query/sql/drivers/odbc/connect/bad_oidc_login",
+                error_description,
+            }
+            .render()
+            .unwrap(),
+        ))
+}
+
+async fn not_found() -> Result<HttpResponse> {
+    Ok(HttpResponse::build(http::StatusCode::NOT_FOUND)
+        .content_type("text/html; charset=utf-8")
+        .body(
+            OIDCNotFoundPage {
+                product_docs_link: "https://www.mongodb.com/docs/atlas/data-federation/query/sql/drivers/odbc/connect",
+                product_docs_name: "Atlas SQL ODBC Driver",
+            }
+            .render()
+            .unwrap(),
+        ))
+}
+
+async fn threaded_run_app(
+    sender: mpsc::Sender<ServerHandle>,
+    oidc_params_sender: mpsc::Sender<StdResult<OidcResponseParams, String>>,
+) -> std::io::Result<()> {
+    // srv is server controller type, `dev::Server`
+    let server = HttpServer::new(move || {
+        let oidc_params_sender1 = oidc_params_sender.clone();
+        let oidc_params_sender2 = oidc_params_sender.clone();
+        App::new()
+            // enable logger
+            .wrap(middleware::Logger::default())
+            .service(
+                web::resource("/callback")
+                    .to(move |r| threaded_callback(oidc_params_sender1.clone(), r)),
```

Comment:
Do you have to clone? Can't you move the ownership/take it at this point?

Reply:
1. The clone here is a simple ref count increment
2. Yes, you have to clone, explaining why requires a deep understanding of how async works :/